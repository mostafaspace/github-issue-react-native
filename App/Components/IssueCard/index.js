import React from 'react';
import {Text, View, Image, FlatList} from 'react-native';
import PropTypes from 'prop-types';
import Styles from './styles';
import moment from 'moment';
import {getContrast} from '../../Helpers';
import Icon from 'react-native-vector-icons/Octicons';

export function IssueCard({issue}) {
    return (
        <View style={Styles.container} testID={'IssueCard'}>
            <View style={Styles.contentContainer}>
                {issue.pull_request ? (
                    <Icon name={'git-pull-request'} style={Styles.titleIcon} />
                ) : (
                    issue.state === 'open' && <Icon name={'issue-opened'} style={Styles.titleIcon} />
                )}
                <Text style={Styles.titleText}>{issue.title}</Text>
            </View>
            {issue.labels && issue.labels.length > 0 && (
                <FlatList
                    data={issue.labels}
                    renderItem={({item, index}) =>
                        item.name && (
                            <View style={[Styles.labelContainer, {backgroundColor: '#' + item.color}]}>
                                <Text
                                    key={`label${index}`}
                                    style={[
                                        Styles.labelText,
                                        {
                                            color: getContrast('#' + item.color),
                                        },
                                    ]}
                                >
                                    {item.name}
                                </Text>
                            </View>
                        )
                    }
                    keyExtractor={(item, index) => index.toString()}
                    horizontal
                    contentContainerStyle={Styles.flatContainer}
                    showsHorizontalScrollIndicator={false}
                />
            )}
            {issue.closed_at ? (
                <Text style={Styles.author}>{`#${issue.number} closed ${moment(issue.closed_at).fromNow()} by ${issue.user.login}`}</Text>
            ) : (
                <Text style={Styles.author}>{`#${issue.number} opened ${moment(issue.created_at).fromNow()} by ${issue.user.login}`}</Text>
            )}

            <View style={Styles.footerContainer}>
                {issue.assignees && issue.assignees.length > 0 ? (
                    <FlatList
                        data={issue.assignees}
                        renderItem={({item, index}) =>
                            item.avatar_url && <Image source={{uri: item.avatar_url}} key={`assignee${index}`} style={Styles.assigneeImg} />
                        }
                        keyExtractor={(item, index) => index.toString()}
                        horizontal
                        contentContainerStyle={Styles.assigneeContainer}
                        showsHorizontalScrollIndicator={false}
                    />
                ) : (
                    <View />
                )}
                {issue.comments ? (
                    <View style={Styles.commentsContainer}>
                        <Text style={Styles.commentText}>{issue.comments}</Text>
                        <Icon name={'comment'} style={Styles.commentIcon} />
                    </View>
                ) : null}
            </View>
        </View>
    );
}

IssueCard.propTypes = {
    issue: PropTypes.object.isRequired,
};
