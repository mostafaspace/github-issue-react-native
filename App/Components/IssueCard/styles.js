import {StyleSheet} from 'react-native';

export default StyleSheet.create({
    container: {
        flex: 1,
        marginBottom: 16,
        marginHorizontal: 8,
        backgroundColor: 'white',
        borderRadius: 16,
        shadowColor: '#000',
        shadowOffset: {
            width: 0,
            height: 2,
        },
        shadowOpacity: 0.23,
        shadowRadius: 2.62,

        elevation: 4,
    },
    contentContainer: {
        flexDirection: 'row',
        padding: 8,
        alignItems: 'center',
    },
    titleText: {
        flex: 1,
        fontWeight: 'bold',
        fontSize: 16,
    },
    titleIcon: {
        fontSize: 24,
        color: 'green',
        marginEnd: 8,
        marginStart: 8,
    },
    flatContainer: {
        paddingHorizontal: 16,
        paddingBottom: 8,
    },
    labelContainer: {
        flex: 1,
        borderRadius: 50,
        paddingHorizontal: 8,
        paddingVertical: 4,
        marginVertical: 4,
        marginEnd: 6,
        shadowColor: '#000',
        shadowOffset: {
            width: 0,
            height: 1,
        },
        shadowOpacity: 0.2,
        shadowRadius: 1.41,

        elevation: 2,
    },
    labelText: {
        fontWeight: 'bold',
        fontSize: 10,
    },
    author: {
        fontSize: 12,
        paddingHorizontal: 18,
    },
    assigneeImg: {
        width: 20,
        height: 20,
        borderRadius: 50,
        marginEnd: 4,
    },
    footerContainer: {
        flex: 1,
        flexDirection: 'row',
        paddingVertical: 8,
        alignItems: 'center',
        justifyContent: 'space-between',
    },
    commentsContainer: {
        flexDirection: 'row',
        paddingHorizontal: 16,
        alignItems: 'center',
    },
    commentText: {
        fontSize: 12,
        paddingEnd: 4,
        lineHeight: 12.4,
    },
    commentIcon: {
        fontSize: 12,
        paddingEnd: 4,
    },
    assigneeContainer: {
        fontSize: 12,
        paddingHorizontal: 16,
    },
});
