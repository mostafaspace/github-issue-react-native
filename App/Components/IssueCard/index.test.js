import React from 'react';
import {render, cleanup} from '@testing-library/react-native';
import {IssueCard} from '../IssueCard';
import renderer from 'react-test-renderer';

const obj = {
    pull_request: null,
    state: 'open',
    labels: [],
    title: 'title',
    closed_at: null,
    created_at: 0,
    number: 1,
    user: {login: 'name'},
    comments: [],
    assignees: [],
}

afterEach(cleanup);
it('renders component', () => {
    render(<IssueCard issue={obj} />);
});

it('IssueCard matches snapshot', () => {
    const tree = renderer.create(<IssueCard issue={obj} />).toJSON()
    expect(tree).toMatchSnapshot();
});
