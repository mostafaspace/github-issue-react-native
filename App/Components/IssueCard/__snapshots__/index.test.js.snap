// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueCard matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 16,
      "elevation": 4,
      "flex": 1,
      "marginBottom": 16,
      "marginHorizontal": 8,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.23,
      "shadowRadius": 2.62,
    }
  }
  testID="IssueCard"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "green",
            "fontSize": 24,
            "marginEnd": 8,
            "marginStart": 8,
          },
          Object {
            "fontFamily": "Octicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      style={
        Object {
          "flex": 1,
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      title
    </Text>
  </View>
  <Text
    style={
      Object {
        "fontSize": 12,
        "paddingHorizontal": 18,
      }
    }
  >
    #1 opened 51 years ago by name
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingVertical": 8,
      }
    }
  >
    <View />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "paddingHorizontal": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "lineHeight": 12.4,
            "paddingEnd": 4,
          }
        }
      />
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Object {
              "fontSize": 12,
              "paddingEnd": 4,
            },
            Object {
              "fontFamily": "Octicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
